api/src/main/java/org/openmrs/Encounter.java

```diff
@@ -518,19 +534,28 @@ public void setConditions(Set<Condition> conditions) {
 	 * @param condition - the condition to add
 	 */
 	public void addCondition(Condition condition) {
-		condition.setEncounter(this);
-		getConditions().add(condition);
+		if (conditions == null) {
+			conditions = new LinkedHashSet<>();
+		}
+
+		if (condition != null) {
+			condition.setEncounter(this);
+			conditions.add(condition);
+		}
 	}
-	
+
 	/**
 	 * Remove the given condition from the set of conditions for this Encounter
 	 *
 	 * @param condition - the condition to remove
 	 */
 	public void removeCondition(Condition condition) {
-		if (conditions != null) {
-			conditions.remove(condition);
-		}
+		
+		Optional.ofNullable(conditions).orElse(new LinkedHashSet<>()).stream().filter(c -> !c.getVoided() && c.getUuid().equals(condition.getUuid())).forEach(c -> {
```

Comment:
Shouldn't we just be explicitly relying on the notion of equality here, e.g., `filter(c -> !c.getVoided && condition.equals(c))`?

Reply:
I know that `condition.equals()` doesn't work that way right now, but I thought it was agreed that we'd proceed with an implementation of `condition.equals()` that only verifies UUID.

Reply:
Done.